### PR TITLE
feat(authentication): support OAuth 2.0 scopes

### LINKED ISSUE #32: feat(authentication): support OAuth 2.0 scopes
**Is your feature request related to a problem? Please describe.**
Firebase Authentication allows to specify additional OAuth 2.0 scopes. This is currently not possible with this plugin.

**Describe the solution you'd like**
Add support for OAuth 2.0 scopes.

### PR DESCRIPTION
**Pull request by nkalupahana**
~This is still WIP since GoogleSignIn 6.2 still isn't out, but since it'll be out very soon and this is a somewhat substantial PR, I figured I would open it now to get an initial review.~ I've tested this for Google and GitHub (so should work for all OAuth) -- so I think the only one not covered by testing on my part is Facebook (all platforms) and Google Play Games (Android).

Close #32

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] The changes have been tested successfully.
- [x] A changeset has been created (`npm run changeset`).

### CODE REVIEW

File: packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/FacebookAuthProviderHandler.java [hunk 1]
```diff
@@ -56,6 +60,18 @@ public void onError(FacebookException exception) {
...
+        if (scopes != null) {
+            try {
+                List<String> scopeList = scopes.toList();
+                scopeList.add("email");
+                scopeList.add("public_profile");
+                this.loginButton.setPermissions(scopeList);
+            } catch (JSONException exception) {
+                Log.e(FirebaseAuthenticationPlugin.TAG, "signIn (applying scopes) failed.", exception);
+            }
+        }
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/FacebookAuthProviderHandler.java:73]: Please move this code to a new method, e.g.:

```java
applySignInConfig(call, loginButton);
```

File: packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/GoogleAuthProviderHandler.java [hunk 2]
```diff
@@ -31,6 +37,25 @@ public GoogleAuthProviderHandler(FirebaseAuthentication pluginImplementation) {
...
+
+                for (String scope : scopeList) {
+                    gsob = gsob.requestScopes(new Scope(scope));
+                }
+
+                mGoogleSignInClient = GoogleSignIn.getClient(pluginImplementation.getPlugin().getActivity(), gsob.build());
+            } catch (JSONException exception) {
+                Log.e(FirebaseAuthenticationPlugin.TAG, "signIn (applying scopes) failed.", exception);
+            }
+        }
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/GoogleAuthProviderHandler.java:57]: Please move this code to a new method, e.g.:

```java
mGoogleSignInClient = buildGoogleSignInClient(scopeList);
```

This method can then also be called from the constructor with an empty list.

File: packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/PlayGamesAuthProviderHandler.java [hunk 3]
```diff
@@ -28,6 +34,25 @@ public PlayGamesAuthProviderHandler(FirebaseAuthentication pluginImplementation)
...
+
+                for (String scope : scopeList) {
+                    gsob = gsob.requestScopes(new Scope(scope));
+                }
+
+                mGoogleSignInClient = GoogleSignIn.getClient(pluginImplementation.getPlugin().getActivity(), gsob.build());
+            } catch (JSONException exception) {
+                Log.e(FirebaseAuthenticationPlugin.TAG, "signIn (applying scopes) failed.", exception);
+            }
+        }
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/PlayGamesAuthProviderHandler.java:54]: Please move this code to a new method like in the `GoogleAuthProviderHandler`.

File: packages/authentication/ios/Plugin/Handlers/GoogleAuthProviderHandler.swift [hunk 4]
```diff
@@ -19,9 +19,10 @@ class GoogleAuthProviderHandler: NSObject {
         guard let clientId = FirebaseApp.app()?.options.clientID else { return }
         let config = GIDConfiguration(clientID: clientId)
         guard let controller = self.pluginImplementation.getPlugin().bridge?.viewController else { return }
+        let scopes = call.getArray("scopes", String.self) ?? []
 
         DispatchQueue.main.async {
-            GIDSignIn.sharedInstance.signIn(with: config, presenting: controller) { user, error in
+            GIDSignIn.sharedInstance.signIn(with: config, presenting: controller, hint: "gmail.com", additionalScopes: scopes, callback: { user, error in
```
Comment (robingenz) [inline packages/authentication/ios/Plugin/Handlers/GoogleAuthProviderHandler.swift:25]: Is this `hint` really necessary? If not, please remove it.

File: packages/authentication/src/definitions.ts [hunk 5]
```diff
@@ -235,6 +235,11 @@ export interface SignInOptions {
    * Configures custom parameters to be passed to the identity provider during the OAuth sign-in flow.
    */
   customParameters?: SignInCustomParameter[];
+
```
Comment (robingenz) [inline packages/authentication/src/definitions.ts:238]: Please remove this empty line.

File: packages/authentication/src/web.ts [hunk 6]
```diff
@@ -121,8 +122,23 @@ export class FirebaseAuthenticationWeb
...
+    options?: SignInOptions,
+  ) {
+    if (typeof provider === 'string') provider = new OAuthProvider(provider);
+    if (options?.scopes) {
+      for (const scope of options.scopes) {
+        provider.addScope(scope);
+      }
+    }
+    return provider;
+  }
```
Comment (robingenz) [inline packages/authentication/src/web.ts:138]: I suggest that we proceed here in the same way as above: Revert the changes in this file and create a new method `applySignInConfig(provider: BaseOAuthProvider, options?: SignInOptions)` which is called after you create the provider in each method.

**Review by robingenz — CHANGES_REQUESTED** [review]
Thanks for your contribution!

**Comment by nkalupahana** [comment]
@robingenz fixes implemented and tested, let me know if you want to see anything else.

File: packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/FacebookAuthProviderHandler.java [hunk 7]
```diff
@@ -58,20 +59,23 @@ public void onError(FacebookException exception) {
         }
     }
 
-    public void signIn(PluginCall call) {
-        savedCall = call;
-        JSArray scopes = call.getArray("scopes");
+    private void applySignInConfig() {
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/FacebookAuthProviderHandler.java:62]:
```suggestion
    private void applySignInConfig(LoginButton loginButton) {
```

Please pass the button instance instead of accessing it directly.

File: packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/GoogleAuthProviderHandler.java [hunk 8]
```diff
@@ -29,33 +31,33 @@ public class GoogleAuthProviderHandler {
...
-    public void signIn(PluginCall call) {
-        JSArray scopes = call.getArray("scopes");
-        // Got scopes? Need to get the client again with these new options.
-        if (scopes != null) {
-            try {
-                List<String> scopeList = scopes.toList();
-                GoogleSignInOptions.Builder gsob = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
-                    .requestIdToken(pluginImplementation.getPlugin().getContext().getString(R.string.default_web_client_id))
-                    .requestEmail();
+    private void buildGoogleSignInClient(@Nullable PluginCall call) {
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/authentication/handlers/GoogleAuthProviderHandler.java:37]: Please move this private method under the last public method.

**Review by robingenz — CHANGES_REQUESTED** [review]
Looks good, just a few final changes.

**Comment by nkalupahana** [comment]
Alright, all set. I'll ping you when GoogleSignIn 6.2 comes out so we can get this merged. Thanks!

**Review by robingenz — APPROVED** [review]

**Comment by robingenz** [comment]
@nkalupahana Please format the files and fix the build error.

**Comment by nkalupahana** [comment]
@robingenz should be all good to go!

**Comment by robingenz** [comment]
@nkalupahana Great work! Thank you!
